mllib/src/main/scala/org/apache/spark/ml/feature/RFormula.scala

```diff
@@ -614,3 +652,80 @@ private object VectorAttributeRewriter extends MLReadable[VectorAttributeRewrite
     }
   }
 }
+
+/**
+ * Utility transformer for adding expressions to dataframe using `expr` spark function
+ *
+ * @param exprsToSelect set of string expressions to be added as a column to the dataframe.
+ *                      The name of the columns will be identical to the expression
+ */
+private class ExprSelector(
```

Comment:
I'm unsure about adding a "hidden" stage and MLReader/MLWriter for this - could you explain more why this is necessary?

Reply:
As you suspect, having an extra hidden stage isn't really essential here. I only added it to have less coupling between RFormula and RFormulaModel classes. This can be very well done without it. 

Roughly, this is what RFormula and RFormulaModel classes are doing;
```scala

class RFormula(..., formula) {
   def fit(df) = {
     val parsedFormula = parse(formula)
     var stages = ArrayBuffer()

     val featureColumns = parsedFormula.terms.map {
         ...
         stages += OneHotEncoder()
         ...
     }
    stages += VectorAssembler(featureColumns)
    val pipeline = Pipeline(stages)
    RFormulaModel(parsedFormula, pipeline)
  }
}

class RFormulaModel(parsedFormula, pipeline) {
  def transform(df) = {
    val withFeatures = pipeline.transform(df)
    transformLabel(withFeatures)
  }
}
```
In order to assemble arithmetic expressions in a feature column with `VectorAssembler`, the dataframe which is transformed by `RFormulaModel` needs to have these columns. One way would be to simply add these transformations inside `RFormulaModel.transform` method,  or another way would be to add a pipelined stage inside `RFormula.fit` method. Seeing that all feature column related transformations are done at `RFormula.fit` method, I chose to add a pipelined stage inside `RFormula.fit` method. 

But indeed, having a transformer for just executing a couple of `expr` functions could be too much. If you think it's unnecessary to add an extra stage, let me know and I'll move it's transformations to `RFormulaModel` class.

Reply:
if the pipeline model is persisted the "hidden" stage will show up, I think? it's good intention for the abstract but probably best not to add to the confusion there also since the code isn't too complicated. if sharing the logic is desired later we could refactor it but likely not as a private stage anyway.

Reply:
Got it, thanks! I removed the hidden stage.